lucet-runtime/lucet-runtime-internals/src/future.rs

```diff
@@ -1,139 +1,194 @@
-        // again.
-        *v.downcast().expect("run_async broke invariant")
-    }
-}
+    /// There is no additional yield to the host when the future passed is immediately ready.
+    /// This behavior is the same as implemented by the `.await` operator.
+    ///
+    /// The future passed is polled from within the guest execution context. If the future is not
+    /// immediately ready, the instance will yield and return control to the async executor.
+    /// Later, once the future is ready to make progress, the async executor will return us to
+    /// the guest context, where Lucet will resume polling the future to completion.
+    ///
+    /// For async hostcalls that may yield to the async executor many times, it's recommended that
+    /// you use `tokio::spawn`, or the equivalent from your async executor, which will spawn the task
+    /// to be run from the host execution context. This avoids the overhead of context switching into
+    /// the guest execution context every time the future needs to make progress.
+    ///
+    /// If `Instance::run_async` is not being used to run the VM, this function will return
+    /// `Err(BlockOnError::NeedsAsyncContext)`.
+    pub fn try_block_on<R>(&self, mut f: impl Future<Output = R>) -> Result<R, BlockOnError> {
+        // We pin the future to the stack (a requirement for being able to poll the future).
+        // By pinning to the stack instead of using `Box::pin`, we avoid heap allocations for
+        // immediately-ready futures.
+        //
+        // SAFETY: We must uphold the invariants of Pin, namely that future does not move until
+        // it is dropped. By overriding the variable named `f`, it is impossible to access f again,
+        // except through the pinned reference.
+        let mut f = unsafe { Pin::new_unchecked(&mut f) };
 
-/// A simple future that yields once. We use this to yield when a runtime bound is reached.
-///
-/// Inspired by Tokio's `yield_now()`.
-struct YieldNow {
-    yielded: bool,
-}
+        loop {
+            // Get the AsyncContext, or die if we aren't async
+            let arc_cx = match &self.instance().state {
+                State::Running {
+                    async_context: Some(cx),
+                } => cx.clone(),
+                State::Running {
+                    async_context: None,
+                } => return Err(BlockOnError::NeedsAsyncContext),
+                _ => unreachable!("Access to vmctx implies instance is Running"),
+            };
 
-impl YieldNow {
-    fn new() -> Self {
-        Self { yielded: false }
-    }
-}
+            // build an std::task::Context
+            let mut cx = Context::from_waker(&arc_cx.waker);
+
+            match f.as_mut().poll(&mut cx) {
+                Poll::Ready(ret) => return Ok(ret),
+                Poll::Pending => {
+                    // The future is pending, so we need to yield to the async executor
```

Comment:
Stylistically I might recommend de-indenting all this to the same level as the `match` by having the `Pending` arm just fall-through, that'd help rightward drift here a bit I think